presto-main-base/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReplaceConditionalApproxDistinct.java

```diff
@@ -0,0 +1,209 @@
+        this.functionResolution = new FunctionResolution(functionAndTypeManager.getFunctionAndTypeResolver());
+    }
+
+    @Override
+    public boolean isEnabled(Session session)
+    {
+        return isOptimizeConditionalApproxDistinctEnabled(session) &&
+                functionResolution.supportsArbitraryFunction();
+    }
+
+    @Override
+    public Pattern<AggregationNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(AggregationNode parent, Captures captures, Context context)
+    {
+        VariableAllocator variableAllocator = context.getVariableAllocator();
+        ProjectNode project = captures.get(SOURCE);
+        Assignments inputs = project.getAssignments();
+        boolean changed = false;
+        Assignments.Builder assignments = Assignments.builder();
+
+        Map<VariableReferenceExpression, AggregationNode.Aggregation> aggregations = new LinkedHashMap<>(parent.getAggregations());
+        for (Entry<VariableReferenceExpression, AggregationNode.Aggregation> entry : parent.getAggregations().entrySet()) {
+            VariableReferenceExpression variable = entry.getKey();
+            AggregationNode.Aggregation aggregation = entry.getValue();
+            RowExpression expr;
+
+            if (!isApproxDistinct(aggregation)) {
+                continue;
+            }
+
+            if (hasIfThenConstantInput(aggregation, inputs)) {
+                SpecialFormExpression ifcondition = (SpecialFormExpression) inputs.get(
+                        (VariableReferenceExpression) aggregation.getArguments().get(0));
+                expr = variableAllocator.newVariable("expr", BIGINT);
+                assignments.put((VariableReferenceExpression) expr, replaceIfExpression(ifcondition));
+            }
+            else {
+                continue;
+            }
+
+            changed = true;
+            aggregations.put(variable, new AggregationNode.Aggregation(
+                    new CallExpression(
+                            aggregation.getCall().getSourceLocation(),
+                            "arbitrary",
+                            functionResolution.arbitraryFunction(BIGINT),
+                            BIGINT,
+                            ImmutableList.of(expr)),
+                    aggregation.getFilter(),
+                    aggregation.getOrderBy(),
+                    aggregation.isDistinct(),
+                    aggregation.getMask()));
+        }
+
+        if (!changed) {
+            return Result.empty();
+        }
+
+        ProjectNode modifiedProject = new ProjectNode(
+                project.getSourceLocation(),
+                project.getId(),
+                project.getSource(),
+                assignments.putAll(inputs).build(),
+                project.getLocality());
+
+        return Result.ofPlanNode(new AggregationNode(
+                parent.getSourceLocation(),
+                parent.getId(),
```

Comment:
whenever we create a new node, generate a new plan node id with the id allocator